File: Data Structures & Algorithms/three-integer-sum/submission-1.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        size = len(nums) - 1

        for i in range(len(nums) - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            
            left, right = i + 1, size
            target = -nums[i]

            while right > left:
                total = nums[left] + nums[right]
    
                if total == target:
                    res.append([nums[i],nums[left],nums[right]])
                    left += 1
                    right -=1

                    while left < right and nums[left] == nums[left - 1]:
                        left += 1
                    while left < right and nums[right] == nums[right + 1]:
                        right -= 1
                
                elif target > total:
                    left += 1
                else:
                    right -= 1

        return res

File: Data Structures & Algorithms/search-2d-matrix/submission-1.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {

        if(matrix == null)
            return false;

        int left = 0;
        int right =  matrix[0].length - 1;
        int end = right;
        int top = 0;
        int bottom = matrix.length - 1;

        int colMid;
        int rowMid;

        while(top <= bottom){
            rowMid = top + (bottom - top)/2;
            if(matrix[rowMid][0] == target || matrix[rowMid][end] == target)
                return true;
            else if(target > matrix[rowMid][0] && target < matrix[rowMid][end]){
                while(left <= right){
                    colMid = left + (right - left)/2;
                    if(matrix[rowMid][colMid] == target)
                        return true;
                    else if(matrix[rowMid][colMid] < target)
                        left = colMid + 1;
                    else
                        right = colMid - 1;
                }
                return false;
            }
            else if(matrix[rowMid][0] > target)
                bottom = rowMid - 1;
            else
                top = rowMid + 1;
        }
        return false;
    }
}
